core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+
+/**
+ * Returns an input stream that reads from this source. Closing the stream will also close this source.
+ */
+public fun Source.asNSInputStream(): NSInputStream = SourceNSInputStream(this)
+
+@OptIn(InternalIoApi::class, UnsafeNumber::class)
+private class SourceNSInputStream(
+    private val source: Source,
+) : NSInputStream(NSData()) {
+
+    private val isClosed: () -> Boolean = when (source) {
+        is RealSource -> source::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    private var error: NSError? = null
+    private var pinnedBuffer: Pinned<ByteArray>? = null
```

Comment:
is it possible not to store pinned value here? If stream will not be closed, it will likely be a memory leak of pinned instance

Reply:
This is a solution I came up with with @swankjesse [in this discussion](https://github.com/square/okio/pull/1123#discussion_r928344672) on the Okio PR. Not pinning the buffer has the potential for the memory address returned to the caller in the `buffer` parameter to become invalid if GC decides to move it. Since leaving a stream open would likely leak other resources as well, I think this is the best we can do with the Kotlin/Objective-C memory interop as it is, unless there are other ideas?

Reply:
Ok, I see, that we can't do anything here to remove pinning.
Though:
1. as stated in documentation, this value is only valid until next read operation - so may be we can reset it on read, if available
2. may be we need to attach `Cleaner` for it? though not sure about overhead and is it really needed or not

Reply:
for 2 I also don't have strong opinion, so up to @fzhinkin do decide I think :)